contributors/design-proposals/sig-cli/apply_v2.md

```diff
@@ -0,0 +1,201 @@
+  - Merge strategy is implicit (not show in the patch), but change the result
+  - Merge keys and strategy are incorrect and buggy for several object fields (e.g. Container.Ports)
+    resulting in counter-intuitive results (and there is no diff command to help users see this
+    beforehand)
+  - Patch hides values on the remote object, which may have hidden interactions with the patch values
+    (e.g. Deployment.Spec.RolloutStrategy)
+- Generating a patch to make a change is more complicated than simply making the change to
+  the object read from the server.  The apply command reads the live object from the server
+  and diffs it against the local configuration.  Building a patch containing special merge
+  directives is a more complicated approach than modifying the object read from the server
+  and updating it with a PUT.
+
+#### New approach
+
+- Read the live object
+- Compare the live object to last-recorded and local files
+- Update the fields on the live object that was read
+- Send a PUT to update the modified object
+- If encountering optimistic lock failure, retry after re-reading the object
+
+#### Trade offs
+
+Good:
+
+- Simpler code to maintain
+- Behaves correctly in more instances
+- Easy to look at the output and determine the final state
+
+Bad:
+
+- Larger request sizes to the server than when sending a patch
+
+### Restructure code into modular components
+
+In the current implementation of apply - parsing and traversing the object trees, diffing the
+contents and generating the patch are logically entangled.  This creates maintenance and
+testing challenges.  We should instead modularize discrete responsibilities - collating the
+object values and updating the target object.
+
+#### Phase 1: Parse last-recorded, local, live objects and collate
+
+Provide a structure that contains the last, local and live value for each field.  This
+will make it easy to walk the a single tree when making decisions about how to update the object.
+
+#### Phase 2: Diff and update objects
+
+Use the visitor pattern to encapsulate how to update each field type for each merge strategy.
+Unit test each visit function.
+
+## Alternatives considered
+
+Continue to use patch for apply.  Continue to evolve the PATCH semantics to be more
+forgiving of version skewed environments.  Add new directives as requirements identified.
+
+## Detailed analysis of structure and impact today
+
+The following PRs constitute the focus of ~6 months of engineering work.  Each of the PRs is very complex
+for the work what it is solving.
+
+### Patterns observed
+
+- PRs frequently closed or deferred because maintainers / reviewers cannot reason about the impact or
+  correctness of the changes
+- Relatively simple changes
+  - are 200+ lines of code
+  - modify dozens of existing locations in the code
+  - are spread across 1000+ lines of existing code
+- Changes that add new directives require updates in multiple locations - create patch + apply patch
+
+### PRs
```

Comment:
I want overlays to work, so I think all of these problems still need to be solved for strategic merge patch.

Reply:
Agreed.  We should talk offline about this, since the requirements for overlays may be a subset of what is need for apply to work.  The semantics of apply are a bit different (though there is a lot of overlap) since apply is merging 3 modified versions of the same object to produce an updated object.  Whereas overlays do work in this context.